File: C#/Shop.cs
public static class Shop
{
    public static void Checkout(Customer customer, Cart cart)
    {
        if (cart.IsEmpty)
            throw new Exception("Cart is empty");

        if (cart.HasExpiredItems())
            throw new Exception("Cart has expired items");

        if (cart.HasOutOfStock())
            throw new Exception("Some products are out of stock");

        var shippingItems = cart.GetShipmentDetails();
        double shippingCost = shippingItems.Count * 10;

        double subtotal = cart.Subtotal;
        double total = subtotal + shippingCost;

        customer.Deduct(total);
        cart.ReduceStock();

        ShippingService.Ship(shippingItems);

        Console.WriteLine("** Checkout receipt **");
        foreach (var item in cart.Items)
        {
            Console.WriteLine($"{item.Quantity}x {item.Product.Name,-12} {item.TotalPrice,6:0}");
        }

        Console.WriteLine("----------------------");
        Console.WriteLine($"{"Subtotal",-12}{subtotal,6:0}");
        Console.WriteLine($"{"Shipping",-12}{shippingCost,6:0}");
        Console.WriteLine($"{"Amount",-12}{total,6:0}");
    }
}

File: Java/Shop.java
import java.util.List;

public class Shop {
    public static void checkout(Customer customer, Cart cart) {
        if (cart.isEmpty())
            throw new RuntimeException("Cart is empty");

        if (cart.hasExpiredItems())
            throw new RuntimeException("Cart has expired items");

        if (cart.hasOutOfStock())
            throw new RuntimeException("Some products are out of stock");

        List<Cart.ShipmentDetail> shippingItems = cart.getShipmentDetails();
        double shippingCost = shippingItems.size() * 15;

        double subtotal = cart.getSubtotal();
        double total = subtotal + shippingCost;

        customer.deduct(total);
        cart.reduceStock();

        ShippingService.ship(shippingItems);

        System.out.println("** Checkout receipt **");
        for (CartItem item : cart.getItems()) {
            System.out.printf("%dx %-12s %6.0f\n", item.getQuantity(), item.getProduct().getName(), item.getTotalPrice());
        }

        System.out.println("----------------------");
        System.out.printf("%-12s%6.0f\n", "Subtotal", subtotal);
        System.out.printf("%-12s%6.0f\n", "Shipping", shippingCost);
        System.out.printf("%-12s%6.0f\n", "Amount", total);
    }
} 